### ISSUE TITLE
[BUG] New rows are created in aquifer_parameters table whenever a well with pumping test data is edited

### ISSUE THREAD

**Issue by acatchpole** [bcgov/gwells#2155]
**Describe the Bug**
Whenever a well that has pumping test/aquifer parameter data is edited, a new row is created in the aquifer_parameters table. If this edit does NOT include any changes to the pumping test data, this row is still created, and it will just have data identical to what already exists. In the change history for this well, along with whatever change was intended, there will also be a change showing the pumping test was changed, with the only alteration being a new testing_number (primary key for the table).

**Expected Behaviour**
When a well is edited and those edits do not include aquifer parameters, there should be no change to the aquifer_parameters table, and no change history entries

**Actual Behaviour**
Aquifer parameters table is showing history when you make any edits to the page.

** Steps To Reproduce**
Steps to reproduce the behaviour:
1. Go to a well with an existing pumping test entry
2. Click on 'edit'
3. Make a change anywhere outside of the pumping test section and click 'Save'
4. Scroll down to the 'Change history' and see that, alongside the intended change, there is a change logged for the pumping test data, showing that the Test Number has been changed.

**Comment by acatchpole** [comment]
********Learnings documentation for #2155********
Related Tickets:
[2070](https://github.com/bcgov/gwells/issues/2070) 
[2122](https://github.com/bcgov/gwells/issues/2122) 


'aquifer_parameters_set' is a data key that is passed whenever a well edit (aka activity submission) is made.
It includes all current pumping test data for that well. Whenever an edit to the well is made, this data is included as part of the payload (along with the other properties listed in SubmissionsHome.vue/formSubmit()/doNotSkip plus whatever change is intended).

This inclusion in doNotSkip was added as part of #2070 and was necessary to prevent a data-loss bug where pumping test data was being "deleted" <actually it just had the well_tag_number (WTN) for that row set to NULL> when unrelated edits were made. Now, that still occurs, but a new row is also created with all the same data and the proper WTN, but with a new testing_number (primary key for aquifer_parameters table)

An issue remained where, in the change history, it still looked like the data was being deleted (pumping test data set to 'None'), and then not showing any subsequent edits to that data. This issue was addressed in #2122 by removing 'testing_number' and 'well_tag_number' from the aquifer_parameters_set payload. Now, instead of well edits resulting in a change history that shows pumping test data set to 'None', we see a more accurate history where a new row is created with all the same data, but with a new testing_number.

This latest fix ensures that the change history always shows all edits. BUT it is still not finished as a new row SHOULD NOT be created when doing an unrelated edit to a well.

*****Behaviour in Database*****
A well has pumping test data added to it. An entry is generated in activity_submissions. Two rows are generated in aquifer_parameters(see Figure 1).
The first row generated has the filing_number the same as the entry in activity_submissions but the well_tag_number set to NULL. The second (latest) row has its well_tag_number but the filing_number set to NULL.
**NOTE: **The most recent entry for a specific well in aquifer_parameters will be the only one that shows the well_tag_number. Multiple entries with the same well_tag_number indicate multiple pumping tests for that well.


![Image](https://github.com/bcgov/gwells/assets/113044739/4b1dff22-b22a-44bd-b5d8-fb15308daa8f)

Figure 1

**That well is edited unrelated to pumping test.** An entry is generated in activity_submissions. In aquifer_parameters, the row with the WTN is DELETED and two new rows are generated as above. 
 
These new rows are, in regard to the non-generated elements, identical to the former last rows (from the images: testing_number 78/77 in Figure 2 = 76/75 in Figure 1)


![Image](https://github.com/bcgov/gwells/assets/113044739/d5edcf94-0c9f-4335-b537-ff71989da0c0)

Figure 2


*****Solution Ideas*****
Remove aquifer_parameters_set from the doNotSkip list and figure out how to deal with the dataloss. Right now, the fix for 2122 must be undone (or have a check for null values) before you can do that, as it modifies values in the aquifer_parameter_set on every save. 

OR

Figure out how to stop updates to the aquifer_parameters table when there are no changes being made.

**Comment by davidclaveau** [comment]
There are a couple of things to note with this bug.

## When it works correctly
When you first create a brand new pumping test, you'll make a POST request on formSubmit in `SubmissionsHome.vue` to the `submissions/staff-edit` endpoint and receive a correct `response` with the first pumping test. This seems to be the only time the functionality works correctly from start to finish. You can see this in the console that the data being sent is being passed with the correct data.

```ts
// SubmissionHome.vue file...
console.log('data before', data.aquifer_parameters_set) // <-- console log for before
ApiService.post(PATH, data).then((response) => {
  console.log('data returned', response.data.aquifer_parameters_set) // <-- console log for after
[...]
```

In fact - _before you refresh the page_ - you can make several changes to this pumping test and it will correctly in the console logs.

### Console logs:
![Image](https://github.com/bcgov/gwells/assets/62899351/57833f3c-e739-4f7f-b921-21e636be7f30)

It will also show in Change History correctly.

### Change history:
![Image](https://github.com/bcgov/gwells/assets/62899351/a86a2f6c-dba8-4138-ba15-9426732c39d8)

Correspondingly, it will show correctly in the `aquifer_parameters` table correctly.

### The`aquifer_parameters` table when working:
![Image](https://github.com/bcgov/gwells/assets/62899351/d09d77ba-109d-4e2c-b5ce-defd43e1bd4c)


## When it doesn't work correctly
To recreate this bug we need to _refresh the page_ after creating the pumping test record. We can then start by submitting any other data.

(For this example, `casing_set` data is sent. `casing_set` uses the same functionality as the above `aquifer_parameters_set` but - unlike `aquifer_parameters_set` - it has consistent success and expected behaviour for updates even after refresh.)

For a better indicator of the issue, also add the following console logs to the `SubmissionsHome.vue` POST request.

```ts
// SubmissionHome.vue file...
console.log('aquifer_parameters_set before', data.aquifer_parameters_set)
console.log('casing_set before', data.casing_set) // <-- add console.log for casing_set now
ApiService.post(PATH, data).then((response) => {
  console.log('aquifer_parameters_set returned', response.data.aquifer_parameters_set)
  console.log('casing_set returned', response.data.casing_set) // <-- add console.log for casing_set now
[...]
```
 
Now when you submit a form with new `casing_set` data, you'll notice a discrepancy in the console with our logs. Where before we were sending our `data.aquifer_parameters_set` array with the two objects and _receiving_ two objects back in the `response.data`, we no longer get that data returned to us. Comparatively, `casing_set` does return the data sent as expected.

### Console logs - first `casing_set`
![Image](https://github.com/bcgov/gwells/assets/62899351/65bc7b73-b2b5-4722-beb5-8b7101b67af8)

This isn't unique to the first casing_set data being set. If you _refresh the page_ again, you can add another casing_set item to see the same issue appear.

### Console logs - second `casing_set`
![Image](https://github.com/bcgov/gwells/assets/62899351/4f0e5c1d-3f10-4f32-9abc-ac034d9ba6de)

And _refreshing the page again_ we can make a change to anything else in the form (like `alternateive_specs_submitted`) and see that `aquifer_parameters_set` isn't returned, but `casing_set` is. This might indicate that there's an issue with updating/saving the `aquifer_parameters_set` in the database with the new `filing_number`.

### Console logs - updating any other field (e.g. `alternative_specs_submitted`)
![Image](https://github.com/bcgov/gwells/assets/62899351/8eca22e6-4a6a-4ccd-9c3f-e8d243145615)

(This also shows that if the page is _refreshed_ every time (after each submit) - and not letting the component re-render - then there won't be dataloss).

If we go back into the database, the `casing` table shows a similar layout to what we had with `aquifer_parameters`. For each save, we see an entry into the table _including_ when we submitted `alternative_spec_submitted` on the refresh and there was no change to `casing_set`.

### The`casing` table always works
![Image](https://github.com/bcgov/gwells/assets/62899351/58257c2c-0a94-4afb-9ebb-b9f62c89cf47)

However, the `aquifer_parameters` table has _only_ deleted and replaced the records with `well_tag_number` values, and skipped over adding any new records with the `filing_number`.

### The`aquifer_parameters` table when NOT working
![Image](https://github.com/bcgov/gwells/assets/62899351/34ba04dc-9aad-4d3a-94f8-10bdb58570eb)

## Dataloss

We see dataloss in the `aquifer_parameter` table when we:

1. Save a pumping test
2. Refresh the page.
3. Make a change to any other form field and Save.
4. Without refreshing the page, make another change and Save.
